[Vigneshwar Hariharan] iP

vigonometry opened the pull request:
# Leo - A CLI-based Productivity System
> Hola, me llamo Leo - _"Hello I am Leo"_

Leo frees your mind to focus on the things you **care about**. It is,

- text-based
- interactive
- ~Fun~ _Super Fun_ and,
- INCREDIBLY FAST to use.

## Features:
- **Three task types: Todo, Event and Deadline**
You can now create Todo, Event and Deadline items by prefacing the task content with todo, event or deadline.
Event and Deadline types require further description: `event ... /from DD/MM/YYYY HHMM /to DD/MM/YYYY HHMM`, `deadline ... /by DD/MM/YYYY HHMM`

- **Added task management functions: list, delete, mark as done and unmark completed tasks**
`list` displays all tasks currently saved in your Task List.
`delete <taskId>` deletes task with taskId in your Task List.
`mark <task Id>` and `unmark <taskId>` can be used to mark a task as completed or incomplete.

- **Added task search functions: find**
`find <keyword>` finds all tasks in your Task List that contains the keyword.

## Setup
1. You can download it from [here](https://github.com/vigonometry/ip/releases/tag/A-Jar)
2. Run it on your terminal using `java -jar leo.jar` (Requires `Java 11` or later).
3. Enter your tasks.

It's as **easy** as 1, 2, 3. 

---
If you're a novice programmer, you can learn Object-Oriented Programming using Java too! Here's the `main` method:

```
public static void main(String[] args) throws LeoTaskException {
        new Leo().start();
}
```

- [ ] If you enjoyed my bot, check this box 😉.

Commit:
Add Gradle support

vigonometry committed:
Add start and print methods to greet, echo and handle exit.

vigonometry committed:
Add taskList, addTask and processRequest to handle add, list and exit.

vigonometry committed:
Add Task class and setDone and resetDone methods to track task completion status

vigonometry committed:
Update task functionality according to sample solution

vigonometry committed:
Add Todo, Deadline and Event functionality to Task class and update processRequest method for new task types.

vigonometry committed:
Fix task representation for events and deadlines.

vigonometry committed:
Fix spacing issue for Events, create test cases and update runtest.sh for Leo.

vigonometry committed:
Create LeoException class for exception handling concerning missing content issues.

vigonometry committed:
Add deleteTask method to handle deletion and create new LeoException subclasses.

vigonometry committed:
Add enums for Task classification.

vigonometry committed:
Add leo task package to manage task list functions.

vigonometry committed:
Update test cases and expected functionality.

vigonometry committed:
Add storage methods for TaskList

vigonometry committed:
Fix FileNotFoundException handling

vigonometry committed:
Add datetime parsing for Event and Deadline types.

vigonometry committed:
Add serialization methods to process LocalDateTime serialization.

vigonometry committed:
Merge branch 'branch-Level-7'

vigonometry committed:
Resolve unintended merge conflicts.

vigonometry committed:
Merge branch 'branch-Level-8'

vigonometry committed:
Remove taskList.ser

Nicklelodeon commented on src/main/java/leo/task/Task.java:
```diff
@@ -0,0 +1,196 @@
+
+    }
+
+    public void resetDone() {
```
The method name resetDone() might be confusing, as it could be interpreted as reset to done. Perhaps setAsUndone() could be used?

vigonometry replied:
Updated method name to `setNotDone()`.

Nicklelodeon commented on src/main/java/leo/task/LeoTaskException.java:
```diff
@@ -0,0 +1,42 @@
+    }
+}
+
+class EmptyFieldException extends LeoTaskException {
+    EmptyFieldException(String msg) {
+        super(msg);
+    }
+
+    EmptyFieldException() {
+        super("Task description cannot be empty but...can you do it on a rainy night in Stoke?\n");
+    }
+}
+
+
+class InvalidCommandException extends LeoTaskException {
+    InvalidCommandException() {
+        super("I'm sorry, I don't know what you want. ¿Que miras bobo?\n");
+    }
+}
+
+class MissingDeadlineException extends EmptyFieldException {
+    MissingDeadlineException() {
+        super("When is it due bruv? I never make predictions, and I never will.\n");
+    }
+}
+
+class MissingTimelineException extends EmptyFieldException {
+    MissingTimelineException() {
+        super("Not this again, indicate a from and to. I'm as happy as I can be—but I have been happier.\n");
+    }
+}
+
+class EmptyDeletionException extends LeoTaskException {
+    EmptyDeletionException() {
+        super("Bruh, why are you trying to delete from an empty task list...\n");
+    }
+}
```
I like the fun chat interactions you have created between the bot and the user, especially the cold rainy night in stoke reference :)) Creating specific exception classes that inherit from LeoTaskException is also a great approach!

Nicklelodeon commented on src/main/java/leo/task/TaskList.java:
```diff
@@ -0,0 +1,93 @@
+                throw new EmptyFieldException();
+            }
+
+            switch (rqDetails[0]) {
+                case "bye":
+                    System.out.println("It was nice talking, see you soon!");
+                    break;
+                case "list":
+                    System.out.println("Here are your tasks you legend: ");
+                    for (int i = 0; i < tasks.size(); i++) {
+                        System.out.printf("%d) %s\n", i + 1, tasks.get(i));
+                    }
+                    break;
+                case "mark":
+                    tasks.get(Integer.parseInt(rqDetails[1]) - 1).setDone();
+                    break;
+                case "unmark":
+                    tasks.get(Integer.parseInt(rqDetails[1]) - 1).resetDone();
+                    break;
+                case "delete":
+                    deleteTask(Integer.parseInt(rqDetails[1]));
+                    break;
+                default:
+                    addTask(request);
+                    break;
+            }
```
Should switch and case have the same level of indentation?

vigonometry replied:
Updated to fit the Java coding standard (they should have the same level of indentation).

Nicklelodeon commented on src/main/java/leo/task/TaskList.java:
```diff
@@ -0,0 +1,93 @@
+        }
+    }
+
+    private void deleteTask(int taskId) throws LeoTaskException {
```
I like how you lower cased the abbreviation in the variable name taskId, as you avoided the common mistake of naming it as TaskID instead!

Nicklelodeon reviewed (commented):
Overall, I felt that your naming conventions for methods and variables were really well done. Most of the coding conventions were adhered to too! I also liked your interesting chat UI interactions between the user and chat bot. 
I believe you could include more OOP to improve readability and abstraction in your code.
Some cosmetic changes were suggested!

vigonometry committed:
Add Storage, Ui and Parser packages.

vigonometry committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

vigonometry committed:
Add gradle support.

francisyzy commented on src/main/java/leo/task/Task.java:
```diff
@@ -0,0 +1,201 @@
+    private String taskDesc;
+    private boolean isDone = false;
+    private Type type;
+    public final static List<String> descCommands = Arrays
```
Hi, I think this variable name can be more descriptive, when I first saw it I thought it was short for describeCommands but it seems like it is short for descriptionCommands.

Was thinking you can put your commands as an enum, with a tag to check if it has any descriptions.
Something like:
```java
public enum Command {
    TODO("todo", true),
    LIST("list", false);

    private final String command;
    private final boolean desc;

    Command(String command, boolean desc) {
        this.command = command;
        this.desc = desc;
    }

    public String getCommand() {
        return command;
    }

    public int getDesc() {
        return desc;
```

vigonometry replied:
This is fantastic, thank you

francisyzy commented on src/main/java/leo/task/Task.java:
```diff
@@ -0,0 +1,201 @@
+        Ui.printDivider();
+    }
+
+    public boolean getDone() {
```
Good variable naming for the `isDone` and their respective methods!

francisyzy commented on src/main/java/leo/task/Task.java:
```diff
@@ -0,0 +1,201 @@
+import leo.ui.Ui;
+
+public class Task implements Serializable {
+    private static final long serialVersionUID = 1L;
```
Hey, I think this needs to be in all caps?

> A constant can be named using ALL_CAPS.

vigonometry replied:
Hi Francis, thanks for your comments. Pertaining to this field, the JavaDocs for implementing a `Serializable` object prescribes the field be named as such.

> The serialization runtime associates with each serializable class a version number, called a serialVersionUID, which is used during deserialization to verify that the sender and receiver of a serialized object have loaded classes for that object that are compatible with respect to serialization...:
 `ANY-ACCESS-MODIFIER static final long serialVersionUID = 42L;`

francisyzy commented on src/main/java/leo/task/Task.java:
```diff
@@ -0,0 +1,201 @@
+    public final static List<String> descCommands = Arrays
+            .asList(new String[] { "todo", "deadline", "event", "mark", "unmark", "delete" });
+
+    public final static List<String> commands = Arrays
```
Similarly, for commands I think it should be all caps?

`Variable 'commands' should be in ALL_CAPS (if it is a constant) or be private (otherwise). [ConstantName]`

vigonometry replied:
Thanks for raising this up. Accidentally changed the access modifiers while implementing more classes.

francisyzy reviewed (commented):
I like it how you named your variables! most of it looks good other than a few minor things.

vigonometry committed:
Update commands' access modifiers.

vigonometry committed:
Add JUnit tests for Task.

vigonometry committed:
Add JavaDocs for public classes.

vigonometry committed:
Add search/find method to TaskList.

vigonometry committed:
Add GUI elements

vigonometry committed:
Revert "Add GUI elements"

This reverts commit 59421cfbf74887ef9cdfd574220410965adefbad.

vigonometry committed:
Add GUI for Leo.

vigonometry committed:
Add exception handling and greeting.

vigonometry committed:
Add help feature and command list.

vigonometry committed:
Update GUI methods

Descriptive commands to Leo have different processing needs compared to non-descriptive commands.

For descriptive fields we need to isolate the content following the command but non-descriptive commands require sole direct processing.

As such, two new methods (processRequestGUI and processDescriptiveRequestedGUI) were added to create this abstraction layer.

vigonometry committed:
Add input assertions.

vigonometry committed:
Update GUI color scheme.

GUI colorscheme has been updated to reflect the Argentinian jersey colors.

This is to keep in line with its personality as the football player Lionel Messi who is from Argentina.

Additionally, font changes have been added to make the overall visual more pleasing to the user.

vigonometry committed:
Add user guide for Leo.

vigonometry committed:
Update build.gradle

vigonometry committed:
Merge pull request #2 from vigonometry/branch-Level-9

Update build.gradle

vigonometry committed:
Add release link.

vigonometry committed:
Add exit handler.

vigonometry committed:
Merge branch 'master' of github.com:vigonometry/ip

vigonometry committed:
Add exception handling for invalid user input.

vigonometry committed:
Update GUI Exception handling.

vigonometry committed:
Update README.md

vigonometry committed:
Update documentation for GUI request methods.

vigonometry committed:
Add assertions to getResponse.

vigonometry committed:
Fix JUnit tests
